File: py-ratelimiter/source_repo/src/limiter.py
from __future__ import annotations

import collections
import time
from dataclasses import dataclass, field
from typing import Deque, Dict, Tuple

from .config import LimiterConfig, Mode


@dataclass
class FixedWindowLimiter:
    cfg: LimiterConfig
    window_start: float = field(default_factory=lambda: 0.0)
    count: int = 0

    def allow(self, ts: float) -> Tuple[bool, int, float]:
        if self.window_start == 0.0:
            self.window_start = ts
        window_end = self.window_start + self.cfg.window_seconds
        if ts >= window_end:
            self.window_start = ts
            self.count = 0
            window_end = self.window_start + self.cfg.window_seconds
        if self.count < self.cfg.max_requests:
            self.count += 1
            return True, self.count, window_end
        return False, self.count, window_end


@dataclass
class SlidingWindowLimiter:
    cfg: LimiterConfig
    events: Deque[float] = field(default_factory=collections.deque)

    def allow(self, ts: float) -> Tuple[bool, int, float]:
        window_start = ts - self.cfg.window_seconds
        while self.events and self.events[0] <= window_start:
            self.events.popleft()
        if len(self.events) < self.cfg.max_requests:
            self.events.append(ts)
            reset = self.events[0] + self.cfg.window_seconds
            return True, len(self.events), reset
        reset = self.events[0] + self.cfg.window_seconds
        return False, len(self.events), reset


class RateLimiterEngine:
    def __init__(self, cfgs: Dict[str, LimiterConfig]) -> None:
        self._limiters: Dict[str, object] = {}
        for name, cfg in cfgs.items():
            if cfg.mode == Mode.FIXED:
                self._limiters[name] = FixedWindowLimiter(cfg)
            else:
                self._limiters[name] = SlidingWindowLimiter(cfg)

    def allow(self, limiter_name: str, ts: float) -> Tuple[bool, int, float]:
        limiter = self._limiters.get(limiter_name)
        if limiter is None:
            # unknown limiter: treat as allowed with dummy window
            return True, 0, ts
        return limiter.allow(ts)

File: py-ratelimiter/golden_repo/cmd/internal/limiter/limiter.go
package limiter

import (
    "time"

    "ratelimiter/internal/config"
)

type FixedWindowLimiter struct {
    Cfg         config.LimiterConfig
    WindowStart float64
    Count       int
}

func NewFixedWindowLimiter(cfg config.LimiterConfig) *FixedWindowLimiter {
    return &FixedWindowLimiter{Cfg: cfg}
}

func (l *FixedWindowLimiter) Allow(ts float64) (bool, int, float64) {
    if l.WindowStart == 0 {
        l.WindowStart = ts
    }
    windowEnd := l.WindowStart + float64(l.Cfg.WindowSeconds)
    if ts >= windowEnd {
        l.WindowStart = ts
        l.Count = 0
        windowEnd = l.WindowStart + float64(l.Cfg.WindowSeconds)
    }
    if l.Count < l.Cfg.MaxRequests {
        l.Count++
        return true, l.Count, windowEnd
    }
    return false, l.Count, windowEnd
}

type SlidingWindowLimiter struct {
    Cfg    config.LimiterConfig
    Events []float64
}

func NewSlidingWindowLimiter(cfg config.LimiterConfig) *SlidingWindowLimiter {
    return &SlidingWindowLimiter{Cfg: cfg, Events: []float64{}}
}

func (l *SlidingWindowLimiter) Allow(ts float64) (bool, int, float64) {
    windowStart := ts - float64(l.Cfg.WindowSeconds)
    // evict old
    i := 0
    for _, v := range l.Events {
        if v > windowStart {
            l.Events[i] = v
            i++
        }
    }
    l.Events = l.Events[:i]
    if len(l.Events) < l.Cfg.MaxRequests {
        l.Events = append(l.Events, ts)
        reset := l.Events[0] + float64(l.Cfg.WindowSeconds)
        return true, len(l.Events), reset
    }
    reset := l.Events[0] + float64(l.Cfg.WindowSeconds)
    return false, len(l.Events), reset
}

type Engine struct {
    limiters map[string]any
}

func NewEngine(cfgs map[string]config.LimiterConfig) *Engine {
    m := make(map[string]any)
    for name, cfg := range cfgs {
        if cfg.Mode == config.ModeFixed {
            m[name] = NewFixedWindowLimiter(cfg)
        } else {
            m[name] = NewSlidingWindowLimiter(cfg)
        }
    }
    return &Engine{limiters: m}
}

func (e *Engine) Allow(name string, ts float64) (bool, int, float64) {
    l, ok := e.limiters[name]
    if !ok {
        return true, 0, ts
    }
    switch v := l.(type) {
    case *FixedWindowLimiter:
        return v.Allow(ts)
    case *SlidingWindowLimiter:
        return v.Allow(ts)
    default:
        return true, 0, ts
    }
}

// helper for tests if needed
func NowSeconds() float64 {
    return float64(time.Now().UnixNano()) / 1e9
}
